File: prolog/p-21.pl
:- ensure_loaded('take-drop').

insert_at(Elem, Xs, N, Ys) :-
    length(Xs, L), L1 is L + 1, N1 is N - 1, N > 0 , N =< L1,
    take(N1, Xs, X1s), drop(N1, Xs,X2s),
    append([X1s, [Elem], X2s], Ys).

File: erlang/p_21.erl
-module(p_21).
-export([insert_at/3]).
-import(take_drop, [take/2, drop/2]).


insert_at(Elem, L, N) when (N > 0) and (N =< length(L) + 1) -> 
    take(L, N - 1) ++ [Elem] ++ drop(L, N - 1).
